packages/e2e-test-utils-playwright/src/page/press-key-with-modifier.ts

```diff
@@ -0,0 +1,138 @@
+/**
+ * External dependencies
+ */
+import { capitalize } from 'lodash';
+import type { Page } from '@playwright/test';
+
+/**
+ * WordPress dependencies
+ */
+import { modifiers, SHIFT, ALT, CTRL } from '@wordpress/keycodes';
+import type { WPKeycodeModifier } from '@wordpress/keycodes';
+
+/**
+ * Internal dependencies
+ */
+import type { PageUtils } from './index';
+
+let clipboardDataHolder: {
+	plainText: string;
+	html: string;
+} = {
+	plainText: '',
+	html: '',
+};
+
+/**
+ * Sets the clipboard data that can be pasted with
+ * `pressKeyWithModifier( 'primary', 'v' )`.
+ *
+ * @param  this
+ * @param  clipboardData
+ * @param  clipboardData.plainText
+ * @param  clipboardData.html
+ */
+export function setClipboardData(
+	this: PageUtils,
+	{ plainText = '', html = '' }: typeof clipboardDataHolder
+) {
+	clipboardDataHolder = {
+		plainText,
+		html,
+	};
+}
+
+async function emulateClipboard( page: Page, type: 'copy' | 'cut' | 'paste' ) {
+	clipboardDataHolder = await page.evaluate(
+		( [ _type, _clipboardData ] ) => {
+			const clipboardDataTransfer = new DataTransfer();
+
+			if ( _type === 'paste' ) {
+				clipboardDataTransfer.setData(
+					'text/plain',
+					_clipboardData.plainText
+				);
+				clipboardDataTransfer.setData(
+					'text/html',
+					_clipboardData.html
+				);
+			} else {
+				const selection = window.getSelection()!;
+				const plainText = selection.toString();
+				let html = plainText;
+				if ( selection.rangeCount ) {
+					const range = selection.getRangeAt( 0 );
+					const fragment = range.cloneContents();
+					html = Array.from( fragment.childNodes )
+						.map( ( node ) =>
+							Object.prototype.hasOwnProperty.call(
+								node,
+								'outerHTML'
+							)
+								? ( node as Element ).outerHTML
+								: node.nodeValue
+						)
+						.join( '' );
+				}
+				clipboardDataTransfer.setData( 'text/plain', plainText );
+				clipboardDataTransfer.setData( 'text/html', html );
+			}
+
+			document.activeElement?.dispatchEvent(
+				new ClipboardEvent( _type, {
+					bubbles: true,
+					cancelable: true,
+					clipboardData: clipboardDataTransfer,
+				} )
+			);
+
+			return {
+				plainText: clipboardDataTransfer.getData( 'text/plain' ),
+				html: clipboardDataTransfer.getData( 'text/html' ),
+			};
+		},
+		[ type, clipboardDataHolder ] as const
+	);
+}
+
+/**
+ * Performs a key press with modifier (Shift, Control, Meta, Alt), where each modifier
+ * is normalized to platform-specific modifier.
+ *
+ * @param  this
+ * @param  modifier
+ * @param  key
+ */
+export async function pressKeyWithModifier(
+	this: PageUtils,
+	modifier: WPKeycodeModifier,
+	key: string
+) {
+	if ( modifier.toLowerCase() === 'primary' && key.toLowerCase() === 'c' ) {
```

Comment:
Same here about the logic being different. Also there are some missing handling like `emulateSelectAll`. I'd expect that if we migrate a util it should preserve the previous logic because right now these util is only used to the `copy/paste` tests. But when we migrate more tests that use `pressKeyWithModifier` with the missing handling cases, it will be really hard to debug.

Reply:
The utils are never meant to be one-to-one migration because Playwright and Puppeteer are still very different in some cases. I didn't include `emulateSelectAll` because Playwright seems to be handling it well implicitly. And we can always add it back if it turns out we still need it.

The current strategy is that if we migrate a test then we should make sure it still works as expected. Developers shouldn't rely on the old util to work seamlessly on Playwright. Instead, they should understand the problem the test is trying to solve, and follow the [best practices](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/README.md#best-practices) when migrating/refactoring it. One of the benefits Playwright brings is that we don't need that many utils anymore, so we should think about minimizing them in the process. I'm not saying it's an easy job though, so yeah, it might be hard to debug for a _migration_. But that's expected, IMO, we're still writing e2e tests in the end, which is also not an easy job.